### ISSUE TITLE
Hiding resources from UI is complex

### ISSUE THREAD

**Issue by dcarrion87** [coder/coder#4552]
I am having difficulty understanding the hiding functionality described here: https://coder.com/docs/coder-oss/latest/templates/resource-metadata#hiding-resources-in-the-ui

Also the example given doesn't seem to add up to what is usually available in a standard template where a deployment is not actually defined? E.g. How would I hide a kubernetes_role_binding following that example when my coder resource is a pod. Is it like this? Doesn't seem to work.

```
resource "coder_metadata" "hide_resource" {
  count = data.coder_workspace.me.start_count
  resource_id = kubernetes_role_binding.coder_workspace[0].id
  hide = true
  item {
    key = "name"
    value = kubernetes_pod.main[0].metadata[0].name
  }
}
```

Does this mean that for every resource I want hidden I need to explicitly define it as opposed to only showing the main pod for connection and hide everything else? I'd also be OK with a collapsed view of supplementary resources so users didn't initially see a long list of resources when they looked at the UI and get confused.

I think I am slowly figuring out how it works but I just thought I'd raise the complexity around this feature / not immediately obvious.

<img width="770" alt="Screen Shot 2022-10-15 at 1 02 07 am" src="https://user-images.githubusercontent.com/2899172/195866160-b03d7e2d-3c3f-449f-864b-23a073d4f31e.png">

**Comment by BrunoQuaresma** [comment]
Hi @dcarrion87, you can use the `hide` option. More info: https://coder.com/docs/coder-oss/latest/templates/resource-metadata#hiding-resources-in-the-ui

**Comment by dcarrion87** [comment]
Hi @BrunoQuaresma I linked that in my initial post and it does not work as described.

**Comment by jsjoeio** [comment]
Hmmm...I tried this just now and am getting an error:
```shell
  Error: Unsupported argument
  An argument named "hide" is not expected here. Did you mean "id"?
```

@dcarrion87 are you seeing the same error?

**Comment by jsjoeio** [comment]
OH! I agree, the docs are a bit misleading. I was thinking you could add `hide = true` to any resource and it would _just work_. Let me try again.

**Comment by jsjoeio** [comment]
So it took me a couple tries, but I finally figured it out. Assuming you have this block in your Terraform file:
```tf
resource "docker_volume" "home_volume" {
  name = "coder-${data.coder_workspace.me.owner}-${data.coder_workspace.me.name}-root"
}
```

To hide it, you create another block like this:
```tf
resource "coder_metadata" "hide_docker_volume" {
  resource_id = docker_volume.home_volume.id
  hide = true
}
```

And then on the Workspace page, it will be hidden like so:
<img width="1712" alt="image" src="https://user-images.githubusercontent.com/3806031/195951942-8a1d4874-794b-4012-8ce7-732dfe014099.png">

> I also don't get why I need an item {} block just to hide an entire resource?

I tried it with and without the item block and it looks like you shouldn't need it.

**Comment by dcarrion87** [comment]
@jsjoeio 

Metadata:

```
resource "coder_metadata" "hide_resource" {
  resource_id = kubernetes_role_binding.coder_workspace[0].id
  hide = true
}
```

Trying to upload:
```
✔ Queued [499ms]
✔ Setting up [61ms]
✔ Adding README.md... [63ms]
✔ Parsing template parameters [130ms]
⧗  Detecting persistent resources 
  terraform environment variable: TF_VAR_venture=<value redacted>
  terraform environment variable: TF_VAR_namespace=<value redacted>
  Terraform 1.3.0
  Error: Insufficient item blocks
  At least 1 "item" blocks are required.
✔ Detecting persistent resources [5167ms]
✘ Cleaning Up [122ms]
template import provision for start: recv import provision: plan terraform: terraform plan: exit status 1
Run 'coder templates push --help' for usage.
```

**Comment by dcarrion87** [comment]
Haha I'm creating quite the comedy central here. I've managed to add a rogue resource that is hidden 😆 

With:

```
resource "coder_metadata" "hide_resource" {
  count = data.coder_workspace.me.start_count
  resource_id = kubernetes_role_binding.coder_workspace[0].id
  hide = true
  item {
    key = "coder_workspace"
  }
  item {
    key = "type"
    value = "kubernetes_role_binding"
  }
}
```

<img width="824" alt="Screen Shot 2022-10-15 at 3 04 19 pm" src="https://user-images.githubusercontent.com/2899172/195967949-85033a6c-573f-42ad-91a0-60425f1e6fa8.png">

**Comment by dcarrion87** [comment]
Nup, I'm leaving this one be for now. Dumping way too many hours into it. Users will have to deal with it for now until the mechanism is simpler to manage. There will be too many resources eventually to manage manually.

**Comment by jsjoeio** [comment]
> Trying to upload it requires the item {} block

Now I'm even more confused because I never got that 🤔 The only thing different is this line:
```tf
resource "coder_metadata" "hide_resource"
```

You called it `hide_resource` and I called it `hide_<resource-name>`. Unsure if that matters though. 

> It's also going to get crazy having to manage the hide=true resources manually across multiple templates

Yeah I can imagine how that could easily become complex and difficult for whoever needs to manage it. Sorry for all the hoop jumping! This is super valuable feedback though so I'm optimistic we can improve it in the future.

**Comment by dcarrion87** [comment]
@jsjoeio sorry about the delay, just needed a good break from this to focus on some other things. 

No go. Screenshot with E.g.:

![Screen Shot 2022-10-23 at 3 00 53 pm](https://user-images.githubusercontent.com/2899172/197373100-a4fcf714-9f5d-4fca-9369-eea48161bb44.png)

Error:

```> Upload "tf/modules/coder-instance/templates/a100"? (yes/no) yes
✔ Queued [564ms]
✔ Setting up [63ms]
✔ Adding README.md... [73ms]
✔ Parsing template parameters [144ms]
⧗  Detecting persistent resources 
  terraform environment variable: TF_VAR_venture=<value redacted>
  terraform environment variable: TF_VAR_namespace=<value redacted>
  Terraform 1.3.0
  Error: Insufficient item blocks
  At least 1 "item" blocks are required.
✔ Detecting persistent resources [5487ms]
✘ Cleaning Up [116ms]
template import provision for start: recv import provision: plan terraform: terraform plan: exit status 1
Run 'coder templates push --help' for usage. 
```

**Comment by dcarrion87** [comment]
@sharkymark sorry to ping you in like this but I noticed in another post you've successfully been able to hide certain Azure resources? What does your terraform look like and did you get the above?

**Comment by bpmct** [comment]
@dcarrion87 You'll just need to add an `item` block :( We should make this not required in the TF provider.

```hcl
resource "coder_metadata" "example"
  resource_id = ...
  item {
    key = "this"
    value = "won't show up"
  }
  hide = true
}
```

**Comment by dcarrion87** [comment]
Hi @bpmct tried that as per previous messages. Does not work as advertised. At least for these resources I'm trying to hide.

**Comment by fevbibb** [comment]
I've been having trouble hiding random_password type resources, it seems terraform sets their id to none, so coder just hides one at random

**Comment by dcarrion87** [comment]
It's starting to get unruly for us now. Really hoping this becomes easier and actually works for all resources in the near future.

**Comment by jsjoeio** [comment]
@dcarrion87 sorry to hear that! Two questions:
1. Are you able to hide resources?
2. What would the ideal API look like?

Or if it's easier for you, I'd be help to schedule a call and chat in detail about this and how we can make it better. joe@coder.com is my email

**Comment by dcarrion87** [comment]
I know UI customisation feature is unbaked but I thought I would try my luck with UI customisations. I'm not even sure how this is possible?


<img width="574" alt="Screen Shot 2022-11-28 at 7 34 29 pm" src="https://user-images.githubusercontent.com/2899172/204231493-21b24664-d91b-42fd-ad8c-603d78b3f943.png">

<img width="1399" alt="Screen Shot 2022-11-28 at 7 33 49 pm" src="https://user-images.githubusercontent.com/2899172/204231519-d02b3140-ab5a-4eac-8683-40783376f134.png">

**Comment by BrunoQuaresma** [comment]
@dcarrion87 I think this issue is related to #5176 but I think hiding resources that have no agents, as you suggested, is a good thing even after having the 5176 issue fixed. What do you think?

**Comment by dcarrion87** [comment]
Hi @BrunoQuaresma that probably makes sense. At least to non admins and by default. I'm not sure if it provides significant value to general users. I could be missing some use cases.

**Comment by BrunoQuaresma** [comment]
Fixed by https://github.com/coder/coder/issues/5176. @dcarrion87 please, let me know if it stills an issue for you.
